tests/distributions/test_lkj.py

```diff
@@ -0,0 +1,76 @@
+from __future__ import absolute_import, division, print_function
+
+import pytest
+import torch
+from torch.distributions import biject_to, transform_to
+from pyro.distributions.lkj import (corr_cholesky_constraint, CorrLCholeskyTransform)
+from tests.common import assert_tensors_equal
+
+
+@pytest.mark.parametrize("value_shape", [(1, 1), (3, 3), (5, 5)])
```

Comment:
Yes, those three tests make sense.  I will add them and address the other two issues you raised in a push tonight.  Thanks!

Reply:
Actually, now that I'm coding this, I'm not so sure about the third test: 

> * For `eta=1., d > 2`, the log_prob of any sample should be a constant, i.e.

This doesn't seem to be correct to me.   If it were correct, its equivalent to saying that the log_prob is determined only by `eta` and `d`, and not by the content of the matrix.   

See also lines 73-78 here: https://github.com/stan-dev/math/blob/983061e8bc4fbbb5bc1bd2d5b029b599c886880b/stan/math/prim/mat/prob/lkj_corr_cholesky_lpdf.hpp#L73

Can you elaborate on what you have in mind?

Reply:
Just briefly looking through their code, it does seem that they special case `eta==1` - `sum(values)` should just be a constant, if I am not wrong.

Reply:
There are three terms in the calculation of the log prob.  One part is the calculation of the lkj constant, which is constant for any given pair of `eta` and `d`.  But there are two other terms.  Look at lines 78-79, 82 (together, the second term) and 85 (the third term) of the code I linked to. 

If eta is 1 **and** not an optimizable parameter, then Stan will skip line 85, since that term drops out.   (This corresponds to line 206 in what I'm about to push.)  

The second term of the calculation takes place at 75-79.  This corresponds to lines 201-204 of the code here.  That part of the calculation is applied even if eta is 1 and a constant; see line 82. 

What do you think?

(Its entirely possible that Stan is deviating from the paper; given the choice, I'd place my bet on Stan's calculation being the correct one.)

Reply:
I have tried to be faithful to the organizational structure of the Stan implementation throughout, to simplify verifying the code's correctness.  I've tried to deviate only to vectorize and avoided Stan's optimizations around dropping constant terms, since Pyro doesn't track which terms are constant, and the optimization benefit would only be obtained by the combination of a strongly-typed language and a strong compiler anyway.

Reply:
@neerajprad the paper derived log_prob of correlation matrix, here @elbamos implements Cholesky version. We need to add jacobian of cholesky transform (of correlation matrix) to make it uniform when eta=1.

Reply:
@fehiepsi The jacobian of the cholesky transformation is the product of the diagonal elements in the cholesky matrix.  I don't see that taking place in the code.  

While the code does take the log-diagonal elements of the cholesky matrix, I'm pretty sure its doing that in order to perform a calculation on the log-determinant of the original matrix -- consistent with @neerajprad's observation that the density is proportional to the determinant^(eta-1).

Reply:
Hmm..it does seem that you wouldn't get a constant density with their implementation, but look at the top of the file with the comment that `eta == 1 <-> uniform`. 😄Maybe my brain has switched off, but I think that it is worth clearing this because the behavior at `eta==1` is called out in many other implementations / descriptions. I'm not too familiar with the Stan code but is it possible that the `lpdf` also does a jacobian adjustment for the transformation?

Reply:
@elbamos I'm not sure that I understand Stan code correctly but it seems to me that the Jacobian adjustment for Cholesky transform is [this term](https://github.com/stan-dev/math/blob/983061e8bc4fbbb5bc1bd2d5b029b599c886880b/stan/math/prim/mat/prob/lkj_corr_cholesky_lpdf.hpp#L79). @neerajprad 's log_prob testing is validate for the case D=2 because jacobian of cholesky transform in that case is 1 (hence, log of it is 0). A way to get that (numerical) Jacobian term is to use `_autograd_log_det` function (in test_lkj.py file). For example
```
# here `sample` is the Cholesky of a correlation matrix
tril_index = sample.new_ones(dimension, dimension).tril(diagonal=-1) > 0.5
# start with tril vector version
sample_tril = sample[tril_index].clone().requires_grad_()
sample_cloned = sample.new_ones(dimension, dimension).tril(diagonal=-1)
sample_cloned_tmp = sample_cloned.clone()
sample_cloned_tmp[tril_index] = sample_tril
sample_cloned_diag = (1 - sample_cloned_tmp.pow(2).sum(-1)).sqrt()
sample_cloned[tril_index] = sample_tril
sample_cloned.view(-1)[::dimension + 1] = sample_cloned_diag
y = sample_cloned.matmul(sample_cloned.transpose(-2, -1))
corr_tril = y[tril_index]
jacobian = _autograd_log_det(corr_tril, sample_tril)
```

Reply:
@fehiepsi I would be more than pleased to include an appropriate test. 

However, in my view, a test needs to satisfy the criteria that it can be implemented without doubt as to both (a) the mathematical validity of the test, and (b) the correctness of the test implementation. 

I would also encourage you to read the entirety of the Stan thread you've linked to. In particular, the part where they discuss the effect of the bug.

Reply:
Yeah, thanks for pointing it out! In HMC, when `eta` is not a parameter, we can remove normalization constant to save computation and to avoid worrying about if it is correctly implemented or not, so it will not affect the inference. That and the lack of tests might be the reason why the bug is lately detected. However, the bug will affect SVI or Important Inference. Anyway, I believe that by adding above tests, I would prefer reading your code than following Stan's code (which I couldn't understand how they derive their terms so I have to rely on Lewandowski et. paper instead :( ). :)

About the math for Cholesky transformation, I write down what I understand here in case it helps
```
# Note about computing Jacobain of the transformation from Cholesky factor to
# correlation matrix:
#
#   Assume C = L@Lt and L = (1 0 0; a \sqrt(1-a^2) 0; b c \sqrt(1-b^2-c^2)),
#   Then off-diagonal lower triangular vector of L is transformed to the off-diagonal
#   lower triangular vector of C by the transform:
#       (a, b, c) -> (a, b, ab + c\sqrt(1-a^2))
#   Hence, Jacobian = 1 * 1 * \sqrt(1 - a^2) = \sqrt(1 - a^2) = L22, where L22
#       is the 2th diagonal element of L
#   Generally, for a D dimensional matrix, we have
#       Jacobian = L22^(D-2) * L33^(D-3) * ... * Ldd^0
```
It seems similar to the term under your log_diagonals's implementation. Though I am not sure if they are exactly the same or not. Maybe some constants else are involved.

Reply:
I'm not sure why you believe it would affect SVI or Importance inference...  We're talking about a constant term.  Its also (by the way) a term that's totally separate from and irrelevant to the discussion we're having.   And we are including tests that would spot that bug, if it existed in this code.

The problem with the tests you propose is that I'm not convinced that they are either mathematically correct or implemented properly.  

To be blunt, I'm not sure what you're trying to get at with these tests -- what they purport to be testing, or why you think its important.  

Let's say the Stan code does include the log abs det jacobian of the cholesky decomposition. That would not imply that the calculation is wrong, or wrongly implemented, either in Stan or in this PR.

Reply:
I don't think that your code is wrong. And I also think that Stan devs have fixed their code. I just suggest a way to test your code. It seems to me that your code will give uniform distribution for the case eta=1 after adjusting Jacobian (because 2(eta-1)... term will be 0). Maybe you get me wrong.

Reply:
I tried it.  Looks good!

Reply:
Nice!